### PR TITLE
Add ability to transform elliptical arcs

### PR DESCRIPTION
**Pull request by alexjlockwood**
Seems to work with some basic SVGs.

I mostly tested using a simple SVG like this:

```
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 200 200">
  <g transform="translate(100 100) rotate(90) scale(2 1.5)">
    <circle r="25" cx="100" cy="100" fill="#000" transform="translate(-100,-100)"/>
  </g>
</svg>
```

Here's a slightly more advanced one:

```
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="-395 197 168 168">
  <use xlink:href="#a" />
  <defs>
    <circle id="a" cx="-311" cy="281" r="84" fill="#04c07d" />
    <path id="b" d="M-268.3 240.7s15.2 13.1-1.6 24.1c0 0-27.2 21-14.9 63.2 0 0 15.1 26.2 14.5 68.8 0 0-.8 14.5-8.3 33.8l-97.5-32.6s14.6-18.9 22.1-41c0 0 18.8-59.2 7.5-84.2 0 0-4.3-2-10.6 1.5 0 0-7.1-7.4 7.2-47 0 0 12.2-26.1 39.9-24.4-.1-.1 30.6 3.5 41.7 37.8z" fill="#7f4200" />
  </defs>
  <g class="st1">
    <use xlink:href="#b" />
    <path class="st3" d="M-280.3 336.3l-71.3 12s-2.9 22.2-24.5 49.6l97.8 32.7c-.1.1 21-45.6-2-94.3z" fill="#fa0" />
    <circle class="st4" cx="-332.8" cy="255.6" r="2.9" fill="#351800" />
    <path class="st5" d="M-268.3 240.7s-15 3.8-31.2 16c0 0-4.3-15.8-13.7-17.8 0 0-7.6-2.9-19.8.6 0 0-13.7 3.5-20.8-.9 0 0 4.7-36.7 43.7-35.7 0-.1 34.2 2.2 41.8 37.8zm-57.2 51.8s-23.6 8.6-26.7-20.2c0 0 14.5-9.7 26.7 20.2z" fill="#333" />
    <path class="st6" d="M-284.6 327.8s-23.4 20.3-63.8 7.3l-17.1 27 86.1-23.5s-.9-6.8-5.2-10.8z" fill="#fff" />
  </g>
</svg>
```

And another one that I can barely comprehend myself just for the lols:

```
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 300 300">
  <g transform="scale(1.4,1.4) translate(30,50)">
    <path d="M0,0 A50,50 0 0,0 25,25" fill="none" stroke="red" stroke-width="2" transform="translate(100,50)"/>
    <path d="M0,0 A50,50 0 1,0 25,25" fill="none" stroke="blue" stroke-width="2" transform="translate(100,50)"/>
  </g>
</svg>
```

### CODE REVIEW

**Comment by alexjlockwood** [comment]
It seems like this PR fixes #73. yay!

**Comment by alexjlockwood** [comment]
This PR also fixes #73, yay!

File: app/components/canvas/canvas.js [hunk 1]
```diff
@@ -309,12 +309,13 @@ class CanvasController {
...
 
-    if (currentArtwork.alpha != 1) {
+    if (currentAlpha != 1) {
       let oldGlobalAlpha = ctx.globalAlpha;
       ctx.globalAlpha = currentAlpha;
+      offscreenCtx.save();
       ctx.scale(1 / this.backingStoreScale_, 1 / this.backingStoreScale_);
       ctx.drawImage(offscreenCtx.canvas, 0, 0);
-      ctx.scale(this.backingStoreScale_, this.backingStoreScale_);
+      offscreenCtx.restore();
```
Comment (romannurik) [inline app/components/canvas/canvas.js]: move this to a separate CL

File: app/scripts/svgpathdata.js [hunk 2]
```diff
@@ -76,28 +76,48 @@ export class SvgPathData {
...
-    let {length, bounds} = computePathLengthAndBounds_(this.commands_);
+    let { length, bounds } = computePathLengthAndBounds_(this.commands_);
     this.length = length;
     this.bounds = bounds;
   }
 
+
   static interpolate(start, end, f) {
-    if (!end || !start || !end.commands || !start.commands
+    if (!end || !start || !end.commands || !start.commands
```
Comment (romannurik) [inline app/scripts/svgpathdata.js]: extra newline

**Review by romannurik — CHANGES_REQUESTED** [review]
awesome!

File: app/components/canvas/canvas.js (hunk 1 above)
Comment (alexjlockwood) [reply app/components/canvas/canvas.js]: Done.

File: app/scripts/svgpathdata.js (hunk 2 above)
Comment (alexjlockwood) [reply app/scripts/svgpathdata.js]: Done.

**Comment by alexjlockwood** [comment]
I addressed your review comments (not sure if there is anything else I need to do... still learning GitHub :p).

File: app/components/canvas/canvas.js [hunk 3]
```diff
@@ -140,11 +140,11 @@ class CanvasController {
     [this.canvas_, this.offscreenCanvas_].forEach(canvas => {
       canvas
           .attr({
-            width: this.artwork.width * this.backingStoreScale_, 
+            width: this.artwork.width * this.backingStoreScale_,
```
Comment (romannurik) [inline app/components/canvas/canvas.js]: stray change in this CL :)

I'd suggest looking here and spotting anything that doesn't belong:
https://github.com/romannurik/AndroidIconAnimator/pull/74/files

File: app/components/canvas/canvas.js [hunk 4]
```diff
@@ -140,11 +140,11 @@ class CanvasController {
...
             height: this.artwork.height * this.backingStoreScale_,
           })
           .css({
-            width: this.artwork.width * this.scale_, 
+            width: this.artwork.width * this.scale_,
```
Comment (romannurik) [inline app/components/canvas/canvas.js]: stray

File: app/components/canvas/canvas.js [hunk 5]
```diff
@@ -309,7 +309,7 @@ class CanvasController {
       drawLayer_(artworkCtx, currentArtwork, true);
     }
 
-    if (currentArtwork.alpha != 1) {
+    if (currentAlpha != 1) {
```
Comment (romannurik) [inline app/components/canvas/canvas.js]: stray

**Review by romannurik — CHANGES_REQUESTED** [review]

File: app/components/canvas/canvas.js (hunk 3 above)
Comment (alexjlockwood) [reply app/components/canvas/canvas.js]: Done.

File: app/components/canvas/canvas.js (hunk 4 above)
Comment (alexjlockwood) [reply app/components/canvas/canvas.js]: Done.

File: app/components/canvas/canvas.js (hunk 5 above)
Comment (alexjlockwood) [reply app/components/canvas/canvas.js]: Done.

**Review by romannurik — APPROVED** [review]
